Sources/SwiftHttp/UrlSessionHttpClient.swift

```diff
@@ -6,81 +6,136 @@
+		} else {
+			res = try await asyncMethod(with: urlRequest, session.downloadTask)
+		}
+#endif
+		guard let pathData = res.0.path.data(using: .utf8) else {
+			throw HttpError.invalidResponse
+		}
+		return try HttpRawResponse((pathData, res.1))
+	}
+	
+	private func asyncMethod<T, S: URLSessionTask>(with urlRequest: URLRequest, _ method: @escaping (URLRequest, @escaping (T?, URLResponse?, Error?) -> Void) -> S) async throws -> (T, URLResponse) {
+		try await withCheckedThrowingContinuation { continuation in
+			method(urlRequest) { t, response, error in
+				if let t = t, let response = response {
+					continuation.resume(returning: (t, response))
+				} else {
+					continuation.resume(throwing: error ?? HttpError.invalidResponse)
+				}
+			}
+			.resume()
+		}
+	}
+}
 
-    ///
-    /// Performs a data task (in memory) HTTP request
-    ///
-    /// - Parameter req: The request object
-    ///
-    /// - Throws: `HttpError` if something was wrong with the request
-    ///
-    /// - Returns: The entire HTTP response
-    ///
-    public func dataTask(_ req: HttpRequest) async throws -> HttpResponse {
-        let urlRequest = req.urlRequest
-        if log {
-            print(urlRequest.curlString)
-        }
-        let res = try await session.data(for: urlRequest)
-        return try HttpRawResponse(res)
-    }
-    
-    ///
-    /// Uploads the contents of the request and returns the response
-    ///
-    /// - Parameter req: The request object
-    ///
-    /// - Throws: `HttpError` if something was wrong with the request
-    ///
-    /// - Returns: The entire HTTP response
-    ///
-    public func uploadTask(_ req: HttpRequest) async throws -> HttpResponse {
-        let urlRequest = req.urlRequest
-        guard let data = urlRequest.httpBody else {
-            throw HttpError.missingUploadData
-        }
-        if log {
-            print(urlRequest.curlString)
-        }
-        let res: (Data, URLResponse) = try await session.upload(for: urlRequest, from: data, delegate: nil)
-        return try HttpRawResponse(res)
-    }
-    
-    ///
-    /// Downloads the contents of the request and returns the response
-    ///
-    /// - Parameter req: The request object
-    ///
-    /// - Throws: `HttpError` if something was wrong with the request
-    ///
-    /// - Returns: The entire response, setting the file location url as an encoded utf8 string as the response data
-    ///
-    public func downloadTask(_ req: HttpRequest) async throws -> HttpResponse {
-        let urlRequest = req.urlRequest
-        if log {
-            print(urlRequest.curlString)
-        }
-        let res: (URL, URLResponse) = try await session.download(for: urlRequest, delegate: nil)
-        guard let pathData = res.0.path.data(using: .utf8) else {
-            throw HttpError.invalidResponse
-        }
-        return try HttpRawResponse((pathData, res.1))
-    }
+extension HttpClient where Self == UrlSessionHttpClient {
```

Comment:
Why do you need this extension? You can init a new client when it is needed... 🤔

Reply:
Static variables are more predictable and convenience, autocomplete works better with it
Apple uses the same approach in SwiftUI (PickerStyle, ListStyle etc)

Reply:
I removed it